src/hooks/useSelector.js

```diff
@@ -12,61 +11,70 @@ function useSelectorWithStoreAndSubscription(
   store,
   contextSub
 ) {
-  const [, forceRender] = useReducer(s => s + 1, 0)
+  const [state, setState] = useState(() => ({
+    storeState: store.getState(),
+    selector,
+    selectedState: selector(store.getState())
+    // subscriptionCallbackError: undefined
+  }))
 
   const subscription = useMemo(() => new Subscription(store, contextSub), [
     store,
     contextSub
   ])
 
-  const latestSubscriptionCallbackError = useRef()
-  const latestSelector = useRef()
-  const latestSelectedState = useRef()
-
-  let selectedState
+  let selectedState = state.selectedState
 
   try {
-    if (
-      selector !== latestSelector.current ||
-      latestSubscriptionCallbackError.current
-    ) {
-      selectedState = selector(store.getState())
-    } else {
-      selectedState = latestSelectedState.current
+    if (selector !== state.selector || state.subscriptionCallbackError) {
+      const newSelectedState = selector(state.storeState)
+      if (!equalityFn(newSelectedState, selectedState)) {
```

Comment:
potential staleness issue here: if the selector changes, but the result does not, the selector does not get updated in the state, i.e. the callback from the effect uses a stale selector

Reply:
Thanks for the comments. Pretty much appreciated.

I knew the staleness, but yeah, this is critical.

```js
state = 3;
selector1 = state => state - 2;
selector2 = state => state / 3;
selector1(state) === 1
selector2(state) === 1
```

Hmm, I don't think I have any solutions to this now, except for stable selectors (useCallback) which is a breaking change.